Envoy policy filter with L3-dependent L7 policy enforcement.

jrajahalme opened the pull request:
Use the new cilium network policy API to configure network policies to Envoy:

- [x] NPDS gRPC server infra.
- [x] Fix pkg/envoy/xds NACK handing.
- [x] Fix NPDS generation to get rid of validation errors on Envoy side.
- [x] Populate searchable data structure from npds protobufs
- [x] Use npds policy for L3-dependent L7 filtering.
- [x] Replace RDS API server with static route configuration
- [x] Make redirect add/update wait for NPDS acks instead of RDS acks.

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -295,20 +300,29 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
+						requestLog.Debug("ACK received but no observers are waiting for ACKs")
 					}
+				} else if versionInfo != nil {
+					requestLog.Debug("NACK received for version ", state.version, " waiting for a version update before sending again")
```
I'd use the Warning level for this one. It will typically indicate a bug in the agent, but one we may be able to recover from.
And use a format string:
```
requestLog.Warningf("NACK received for version %d; waiting for a version update before sending again", state.version)
```

rlenglet commented on pkg/envoy/xds/server.go:
```diff
@@ -295,20 +300,29 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
-				} else if versionInfo != nil {
-					// If no pending watch exists, then this request is an ACK
-					// for the last response for this resource type.
+				if versionInfo != nil && *versionInfo == state.version {
```
How about:
```
if versionInfo != nil {
    if *versionInfo == state.version {
        ... // ACK
    } else {
        ... // NACK
    }
}
```

rlenglet reviewed (commented):
Reviewed the NACK handling only for now. LGTM. Just a few nitpicks.

rlenglet commented on envoy/cilium/cilium_bpf_metadata.proto:
```diff
@@ -10,6 +10,6 @@ message BpfMetadata {
-  // Cilium security identity for which this proxy is confugured.
-  uint32 identity = 3;
+  // Cilium endpoint security policy to enforce
+  string policy_name = 3;
```
Is the name of this message (`BpfMetadata`) still relevant? This field has nothing to do with BPF.

jrajahalme replied:
I'll move the policy name configuration to the L7 filter instead.

rlenglet commented on envoy/cilium/cilium_l7policy.proto:
```diff
@@ -3,10 +3,15 @@ syntax = "proto3";
   string access_log_path = 1;
 
   // Cilium identifier for the listener.
   string listener_id = 2;
```
If I understand correctly, this is what is passed back in access log entries and used to correlate back to Listeners. This won't work with sidecar proxies. We only need 2 pieces of information for access logging:

-  the identity of the endpoint,
- the direction (ingress / egress)

Both are already given in `BpfMetadata`. How about we remove this field, and just pass in access log entries the values of the `is_ingress` and `policy_name` fields from `BpfMetadata`?

We can do it after this PR is merged.

jrajahalme replied:
Postponing this till after.

rlenglet commented on envoy/cilium_l7policy.h:
```diff
@@ -54,7 +59,8 @@ typedef std::shared_ptr<Config> ConfigSharedPtr;
 // the shared configuration.
-class AccessFilter : public Http::StreamFilter {
+class AccessFilter : public Http::StreamFilter,
+                     Logger::Loggable<Logger::Id::router> {
```
router -> filter

jrajahalme replied:
Done.

rlenglet commented on envoy/cilium_l7policy.h:
```diff
@@ -33,17 +35,20 @@ struct FilterStats {
  * is accessed by multiple working thread instances of the filter.
  */
-class Config : Logger::Loggable<Logger::Id::router> {
+class Config : public Logger::Loggable<Logger::Id::router> {
```
router -> filter

jrajahalme replied:
Done.

rlenglet commented on envoy/cilium_network_policy.cc:
```diff
@@ -0,0 +1,115 @@
+	npmap.erase(policy_name);
+      }
+      for (const auto& new_policy: *to_be_added) {
+	ENVOY_LOG(debug, "Cilium inserting network policy for endpoint {}", new_policy->policy_proto_.name());
```
"inserting" -> "updating", or "upserting", or "setting"?

jrajahalme replied:
Went with "updating"

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+	  ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRule(): Allowing remote {}", remote);
+	  allowed_remotes_.emplace(remote);
+	}
+	for (const auto& http_rule: rules.http_rules().http_rules()) {
```
You must check that `rules.http_rules()` is set before doing that loop.
This can be the case when mixing L4-only rules and L4+L7 rules in the same policy.

jrajahalme replied:
Thanks! Added check for `rules.has_http_rules()` before the loop.

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+
+      bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	// Remote ID must match if we have any.
+	if (allowed_remotes_.size()) {
```
To be consistent in style with other checks below:
```
if (allowed_remotes_.size() > 0) {
```

jrajahalme replied:
Done.

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+	  }
+	}
+	// Empty set matches any payload
+	if (http_rules_.size() == 0) {
```
I would restructure this check to be similar to the check above:
```
if (http_rules_.size() > 0) {
  bool matches = false;
  for (const auto& rule: http_rules_) {
    if (rule.Matches(headers)) {
      matches = true;
      break;
    }
  }
  if (!matches) {
    return false;
  }
}
return true;
```

jrajahalme replied:
As there is no need for the variable here, I wrote this as:
```
	if (http_rules_.size() > 0) {
	  for (const auto& rule: http_rules_) {
	    if (rule.Matches(headers)) {
	      return true;
	    }
	  }
	  return false;
	}
	// Empty set matches any payload
	return true;
```

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+    public:
+      PortNetworkPolicy(const google::protobuf::RepeatedPtrField<cilium::PortNetworkPolicy>& rules) {
+	for (const auto& it: rules) {
+	  // TODO: Handle Missing Protocol
```
I think there's nothing to do.
I would just comment that we ignore UDP rules because they are not relevant to an HTTP filter.

jrajahalme replied:
Done.

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+	  if (it.protocol() == envoy::api::v2::core::SocketAddress::TCP) {
+	    // port may be zero, which matches any port.
+	    ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicy(): installing TCP policy for port {}", it.port());
+	    rules_.emplace(it.port(), PortNetworkPolicyRules(it.rules()));
```
Would it be useful to add a check that there is no PortNetworkPolicyRules associated with this TCP port yet?
The Cilium agent ensures that, but it may be good to add a check here?

jrajahalme replied:
I added check for the return value's `second` member which tells if a new element was added. If not, an exception is thrown, which will result in a xDS NACK.

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+      }
+
+      bool Matches(uint32_t port, uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	auto range = rules_.equal_range(port);
```
I don't understand why you use `equal_range` instead of `find`. There can be at most one entry in the map with that port number, right?

jrajahalme replied:
Remnant from the time I was planning to support duplicates, will change to a simple `find()`.

rlenglet commented on envoy/cilium_network_policy.h:
```diff
@@ -0,0 +1,218 @@
+	    ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicy(): installing TCP policy for port {}", it.port());
+	    rules_.emplace(it.port(), PortNetworkPolicyRules(it.rules()));
+	  } else {
+	    ENVOY_LOG(warn, "Cilium L7 PortNetworkPolicy(): NOT installing non-TCP policy");
```
Reduce this to debug.

jrajahalme replied:
Done.

rlenglet reviewed (approved):
The logic in the filter LGTM.

ianvernon reviewed (approved):
Approving for CI.

eloycoto commented on test/Vagrantfile:
```diff
@@ -10,7 +10,7 @@ $SERVER_BOX= "cilium/ubuntu"
 
 # RAM and CPU settings
-$MEMORY = (ENV['MEMORY'] || "2048").to_i
+$MEMORY = (ENV['MEMORY'] || "4096").to_i
```
I would suggest not change the default value to 4GB here. If you need 4GB in your scenario, you can use the ENV variable (I'm using direnv for this things). If you want to have 4GB in Jenkins, Jenkinsfile is a better place. 
Example: https://github.com/cilium/cilium/blob/master/ginkgo-kubernetes-all.Jenkinsfile#L16

jrajahalme replied:
There are three of the ginkgo Jenkinsfiles. Can you tell me which of those will do "make" to build Cilium?

raybejjani replied:
Does this also mean that the dev VM needs more memory too?

jrajahalme committed:
Ginkgo: Upgrade default memory from 2GB to 4GB.

CI bazel caches are not fresh for When rebasing Envoy, add more memory
to be able to compile the C++ files.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Clean up locking.

Consumable needs to be locked and PolicyMap to exist when updating network policy cache.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Call OnProxyPolicyUpdate in a goroutine.

This should get around a deadlock experienced in testing.

Suggested-by: Romain Lenglet <romain@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

raybejjani commented on pkg/envoy/server.go:
```diff
@@ -534,7 +470,7 @@ func UpdateNetworkPolicy(ep NetworkPolicyEndpoint, policy *policy.L4Policy,
 		policyRevision := policy.Revision
 		callback := func() {
-			ep.OnProxyPolicyUpdate(policyRevision)
+			go ep.OnProxyPolicyUpdate(policyRevision)
```
Is it worth adding a comment explaining why this starts a goroutine instead of doing it synchronously?

raybejjani commented on pkg/proxy/envoyproxy.go:
```diff
@@ -53,7 +59,15 @@ func createEnvoyRedirect(r *Redirect, wg *completion.WaitGroup) (RedirectImpleme
+			return nil, fmt.Errorf("%s: Cannot create redirect, proxy source has no IP address.", r.id)
+		}
+		envoyProxy.AddListener(r.id, ip, r.ProxyPort, r.ingress, redir, wg)
+		atomic.AddInt32(&redirectCount, 1)
```
Since this variable is accessed across files, it might be worth wrapping the increment/get with functions. On the other hand, it's very clear what's going on now. shrug.

raybejjani replied:
an alternative would be to make it a custom type, and making them methods. Still more complexity, however

rlenglet replied:
Agreed. I plan to make it a private field in XDSServer in a subsequent PR.

jrajahalme committed:
ginkgo: Disable flakey runtime/ct.go.

Some of these tests randomly fail, so disable them for now.

Suggested-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Replace RDS routes with a default route.

Now that policy enforcement is done with NPDS, routing can be separated from policy enforcement.

Install a default route to route all traffic to the original destination cluster.

Decide to wait or not based on an atomic count of redirects.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: Romain Lenglet <romain@covalent.io>

jrajahalme committed:
envoy: Add gRPC Network Policy client streamer and L3-dependent L7 enforcement.

Use Cilium Network Policy Discovery Service (NPDS) to sync Cilium
network policy to Cilium Envoy filters. Enforce L3-dependent L7
policies.

Use file-based subscription for Cilium Network Policy for testing without gRPC.

Remove updates from envoy_test.go, as Envoy policy is no longer
updated vie the redirect API.

Update to the new builder with updated Bazel caches.
Update docker-compose.yml to cilium-builder:2018-03-16.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
